File: src/types/musicAnalysis.ts
export type MusicAnalysisStatus =
  | 'ready'
  | 'experimental'
  | 'unavailable'
  | 'failed'


export type ChordEvent = {
  chord: string

  measure:
    number | null

  beat:
    number | null

  time_seconds:
    number | null

  confidence:
    number | null
}


export type MusicAnalysis = {
  song_id: string

  provider: string

  status:
    MusicAnalysisStatus

  tempo_bpm:
    number | null

  key:
    string | null

  time_signature:
    string | null

  chords:
    string[]

  chord_events:
    ChordEvent[]

  confidence:
    number | null

  message:
    string | null
}

File: backend/app/schemas/music_analysis.py
from enum import Enum

from pydantic import (
    BaseModel,
    Field,
)


class MusicAnalysisStatus(
    str,
    Enum,
):
    ready = "ready"

    experimental = (
        "experimental"
    )

    unavailable = (
        "unavailable"
    )

    failed = "failed"


class ChordEvent(BaseModel):
    chord: str

    measure: int | None = Field(
        default=None,
        ge=1,
    )

    beat: float | None = Field(
        default=None,
        ge=1,
    )

    time_seconds: float | None = Field(
        default=None,
        ge=0,
    )

    confidence: float | None = Field(
        default=None,
        ge=0,
        le=1,
    )


class MusicAnalysisResult(BaseModel):
    song_id: str

    provider: str

    status: MusicAnalysisStatus

    tempo_bpm: float | None = None

    key: str | None = None

    time_signature: str | None = None

    chords: list[str] = Field(
        default_factory=list
    )

    chord_events: list[
        ChordEvent
    ] = Field(
        default_factory=list
    )

    confidence: float | None = Field(
        default=None,
        ge=0,
        le=1,
    )

    message: str | None = None